Add availability counters.

wt opened the pull request.

cep21 commented on web/availabilitycounter.go:
```diff
@@ -0,0 +1,36 @@
+	timeKeeper               timekeeper.TimeKeeper
+}
+
+// ModStats ....
```
Can you expand this comment?

cep21 commented on web/availabilitycounter.go:
```diff
@@ -0,0 +1,36 @@
+// ModStats ....
+func (h *AvailabilityCounterHandler) ModStats(ctx context.Context) {
+	start := RequestTime(ctx)
+	fmt.Println(start)
```
Remove println

cep21 commented on web/availabilitycounter_test.go:
```diff
@@ -0,0 +1,52 @@
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+type limitStub struct {
```
When it's a single variable, people generally do
```
type limitStub time.Duration
```

cep21 commented on web/availabilitycounter_test.go:
```diff
@@ -0,0 +1,52 @@
+func TestAvailabilityCounter(t *testing.T) {
+	starttime := time.Date(1981, time.March, 19, 6, 0, 0, 0, time.UTC)
+	fastRequestLimit := 100 * time.Millisecond
+	limitStub := limitStub{
```
Then this becomes
```
stub := limitStub(fastRequestLimit)
```

cep21 commented on web/availabilitycounter_test.go:
```diff
@@ -0,0 +1,52 @@
+			timeKeeper:               timeStub,
+		}
+		ctx := context.Background()
+		Convey("When request is slow", func() {
```
The Convey together should read as a sentence.  So make this one "slow requests should count to the slow bucket" so that the total sentence becomes "When setup, slow requests should count to the slow bucket"

cep21 commented on web/availabilitycounter_test.go:
```diff
@@ -0,0 +1,52 @@
+		timeStub := timekeepertest.NewStubClock(starttime)
+		counterHandler := AvailabilityCounterHandler{
+			fastRequestLimitDuration: limitStub,
+			totalFast:                0,
```
It's OK to skip the zero initialization, especially in tests.

cep21 commented on web/availabilitycounter_test.go:
```diff
@@ -0,0 +1,52 @@
+}
+
+func TestAvailabilityCounter(t *testing.T) {
+	starttime := time.Date(1981, time.March, 19, 6, 0, 0, 0, time.UTC)
```
If you don't actually care what this time is, it's ok to use ```starttime := time.Now()```

cep21 commented on web/availabilitycounter_test.go:
```diff
@@ -0,0 +1,52 @@
+			So(counterHandler.totalSlow, ShouldEqual, 1)
+			So(counterHandler.totalFast, ShouldEqual, 0)
+		})
+		Convey("When request is fast", func() {
```
change this string too

cep21 commented:
Looks great.  Just a few comments.  Can you make a commit on top?  If it looks good I'll squash/merge.

wt commented:
@cep21 Okay, I just realized that the commit I pushed was the wrong one. It from before out discussion yesterday. I think I have addressed all the comments you made in this new commit. However, I pushed a new commit because the last one was just wrong. PTAL.

cep21 commented:
I'll look soon.

When you --force push a new commit, I can't review the delta between the two versions, which really sucks on long commits.

wt commented:
I know. However, the diff would have been pretty big anyway since most of the code was shifted around. Also, I did add another commit to achieve 100% coverage.

cep21 commented on web/reqlatencycounter.go:
```diff
@@ -0,0 +1,63 @@
+
+// ReqLatencyCounter hold static data for the request latency tracking.
+type ReqLatencyCounter struct {
+	fastRequestLimitDuration FastRequestLimitDuration
```
Will probably be difficult to use by others if ```fastRequestLimitDuration``` is private unless you provide a constructor.

wt commented:
PTAL. I added the constructor.

cep21 commented:
I've already merged
